### PR TITLE
255 GitHub actions migration

### LINKED ISSUE #255: Convert from circleci to github actions
### Problem

Our circleci credits are running out.

### Solution

Move to github actions.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by statop**
**Description**:
Migrate to github actions

**Related issue(s)**:

Fixes #255

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: .github/workflows/openjdk_14_build.yml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                              {
+                                  "type": "mrkdwn",
+                                  "text": ":rotating_light: Please correct the issues identified by SonarCloud before merging!"
+                              }
+                          ]
+                      }
+                  ]
+            }
+        env:
+          SLACK_BOT_TOKEN: ${{ secrets.SLACK_ACCESS_TOKEN }}
```
Comment (statop) [inline .github/workflows/openjdk_14_build.yml:140]: the token from circleci doesn't seem to be working, I don't think anyone is watching for these notifications though...

Comment (DSMatusevich) [reply .github/workflows/openjdk_14_build.yml:140]: I didn't even remember there was a slack option. TBH since I was the only one looking at CircleCI, it didn't really matter. WE might want to fix it for the future

File: .github/workflows/openjdk_14_build.yml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+          SONAR_QG_SCRIPT="${REPO_PATH}/.github/scripts/sonar_check_quality_gate.pl";
+          if [[ ! -f "${SONAR_QG_SCRIPT}" ]]; then exit 38; fi;
+
+          SONAR_REPORT_FILE="${REPO_PATH}/target/sonar/report-task.txt";
+          if [[ ! -f "${SONAR_REPORT_FILE}" ]]; then exit 40; fi;
+
+          export SONAR_TASK_ID="$(cat "${SONAR_REPORT_FILE}" | perl -0777 -e 'while (my $line = <>) { if ($line =~ /ceTaskId=([A-Za-z0-9_-]+)/ig) { print "$1"; } }')";
+          if [[ -z "${SONAR_TASK_ID}" ]]; then exit 41; fi;
+          
+          perl ${SONAR_QG_SCRIPT};
```
Comment (steven-sheehy) [inline .github/workflows/openjdk_14_build.yml:62]: Why this very complicated perl script instead of the standard sonar maven plugin? The [docs](https://sonarcloud.io/project/configuration?analysisMode=GitHubActions) suggest this when setting up a project. Mirror node workflow example: https://github.com/hashgraph/hedera-mirror-node/blob/f3848377a55d953b7529633585dc2f2cc65c4490/.github/workflows/security.yml#L53

Comment (statop) [reply .github/workflows/openjdk_14_build.yml:62]: Just copy-pasta from circleci config. I’ll see if I can replace.

File: .github/workflows/openjdk_14_build.yml (hunk 1 above)
Comment (statop) [reply .github/workflows/openjdk_14_build.yml:140]: I’ll set this up properly without custom payload. Who would I talk to about getting a new slack token?

Comment (statop) [reply .github/workflows/openjdk_14_build.yml:106]: done

File: .github/workflows/openjdk_14_build.yml (hunk 2 above)
Comment (statop) [reply .github/workflows/openjdk_14_build.yml]: done

**Comment by steven-sheehy** [comment]
You should add the SonarCloud GitHub app to the project so that it can comment in PRs with its report.

File: .github/workflows/openjdk_14_build.yml (hunk 2 above)
Comment (steven-sheehy) [reply .github/workflows/openjdk_14_build.yml]: You can also specify some of the sonar parameters via the [pom.xml](https://github.com/hashgraph/hedera-mirror-node/blob/e10b4579fb25dd1c02a02f64c5b151c07aca38fd/pom.xml#L84)

Comment (statop) [reply .github/workflows/openjdk_14_build.yml]: they were already there :-)

**Comment by statop** [comment]
@steven-sheehy I think I need an admin of the org level to update settings on either github or sonarqube to do that - https://docs.sonarqube.org/latest/analysis/github-integration/

I added the needed stuff to the pipeline and it doesn't seem to be connecting.

**Comment by steven-sheehy** [comment]
Can reach out to Mark in IT for that.

**Comment by statop** [comment]
I think this is ready for merge.

File: .github/workflows/openjdk_14_build.yml [hunk 3]
```diff
@@ -0,0 +1,117 @@
+name: openjdk_14_build
+on:
+  pull_request:
+  push:
+    branches: [ main, develop ]
+    tags: [ v* ]
+jobs:
+  build-and-test:
+    runs-on: ubuntu-latest
+    container: gcr.io/swirlds-registry/cci-openjdk-ghr:14.0.2
```
Comment (steven-sheehy) [inline .github/workflows/openjdk_14_build.yml:10]: Why the hardcoded swirlds container? We should omit this and use the standard `actions/setup-java@v2` as a step instead.

File: .github/workflows/openjdk_14_build.yml [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    env:
+      MAVEN_OPTS: -Xmx5200m
+      LC_ALL: C.UTF-8
+      DEBIAN_FRONTEND: noninteractive
+    timeout-minutes: 60
+    steps:
+      - run: apt-get -y install software-properties-common
+      - run: add-apt-repository ppa:git-core/ppa
+      - run: apt-get -y install git
```
Comment (steven-sheehy) [inline .github/workflows/openjdk_14_build.yml:19]: I don't think git is necessary to install as it should be there by default for use by the checkout action. What is `software-properties-common` for and is it necessary?

File: .github/workflows/openjdk_14_build.yml [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
```
Comment (steven-sheehy) [inline .github/workflows/openjdk_14_build.yml:22]: Why can't we use a faster shallow clone by omitting this?

File: .github/workflows/openjdk_14_build.yml [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+      - run: mvn --version
```
Comment (steven-sheehy) [inline .github/workflows/openjdk_14_build.yml:23]: This command looks like leftover debug output. You can always add this flag to one of the below mvn commands if it's useful, otherwise remove.

File: .github/workflows/openjdk_14_build.yml [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        with:
+          path: ~/.sonar/cache
+          key: ${{ runner.os }}-sonar
+          restore-keys: ${{ runner.os }}-sonar
+      - name: Cache Maven packages
+        uses: actions/cache@v1
+        with:
+          path: ~/.m2
+          key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
+          restore-keys: ${{ runner.os }}-m2
```
Comment (steven-sheehy) [inline .github/workflows/openjdk_14_build.yml:35]: Use `actions/setup-java@v2` and it has caching built-in:

```suggestion
      - name: Install JDK
        uses: actions/setup-java@v2
        with:
          cache: 'maven'
          distribution: 'temurin'
          java-version: 14
```

File: .github/workflows/openjdk_14_build.yml [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+        run: mvn test verify org.sonarsource.scanner.maven:sonar-maven-plugin:sonar -Dsonar.pullrequest.github.repository=hashgraph/hedera-transaction-tool-demo -Dsonar.pullrequest.provider=github -Dsonar.qualitygate.wait=true -Dsonar.projectKey=com.hedera.hashgraph:hedera-transaction-tool
+      - name: Publish Unit Test Results
+        uses: EnricoMi/publish-unit-test-result-action@v1
+        if: always()
+        with:
+          files: "*/target/surefire-reports/*.xml"
+      - name: Post to a Slack channel
+        uses: slackapi/slack-github-action@v1.16.0
+        if: always()
```
Comment (steven-sheehy) [inline .github/workflows/openjdk_14_build.yml:50]: As this is a public open source repo, we should ensure steps that require secrets can still run for contributors. You should be able to adjust this `if` conditional to see if the secret is present before executing the slack action.

File: .github/workflows/openjdk_14_build.yml [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+          path: ~/.m2
+          key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
+          restore-keys: ${{ runner.os }}-m2
+      - name: Compile
+        run: mvn -DskipTests clean compile
+      - name: Unit Tests And Sonar
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+        run: mvn test verify org.sonarsource.scanner.maven:sonar-maven-plugin:sonar -Dsonar.pullrequest.github.repository=hashgraph/hedera-transaction-tool-demo -Dsonar.pullrequest.provider=github -Dsonar.qualitygate.wait=true -Dsonar.projectKey=com.hedera.hashgraph:hedera-transaction-tool
```
Comment (steven-sheehy) [inline .github/workflows/openjdk_14_build.yml:42]: We should split the test and sonar steps so that we can conditionally run the sonar if the PR contributor has permission for the secrets.

File: .github/workflows/openjdk_14_build.yml [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+          key: ${{ runner.os }}-sonar
+          restore-keys: ${{ runner.os }}-sonar
+      - name: Cache Maven packages
+        uses: actions/cache@v1
+        with:
+          path: ~/.m2
+          key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
+          restore-keys: ${{ runner.os }}-m2
+      - name: Compile
+        run: mvn -DskipTests clean compile
```
Comment (steven-sheehy) [inline .github/workflows/openjdk_14_build.yml:37]: Would highly suggest installing the [Maven Wrapper](https://maven.apache.org/wrapper/) into your repository so it provides repeatable builds. It guarantees CI and local executions are on the same maven version and you can provide JVM and maven configuration in dedicated source controlled [files](https://github.com/hashgraph/hedera-mirror-node/tree/main/.mvn) instead of as `MAVEN_OPTS: -Xmx5200m`. You would then just execute `./mvnw` instead of `mvn` everywhere.

Comment (steven-sheehy) [inline .github/workflows/openjdk_14_build.yml:37]: Add flags helpful in CI like `--batch-mode --no-transfer-progress --show-version`. If you use Maven Wrapper you can add those to the `maven.config`.

File: tools-ui/src/test/java/com/hedera/hashgraph/client/ui/utilities/KeyPairUtilityTest.java [hunk 11]
```diff
@@ -144,7 +144,7 @@ public void mapExpiration_test() {
 
 		homePanePage.enterPasswordInPopup("123654789");
 
-		clickOn("ADD SIGNATURE");
+		clickOn("ADD MORE SIGNATURES");
```
Comment (steven-sheehy) [inline tools-ui/src/test/java/com/hedera/hashgraph/client/ui/utilities/KeyPairUtilityTest.java:147]: This change doesn't seem related to GH actions?

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: .github/workflows/openjdk_14_build.yml (hunk 3 above)
Comment (statop) [reply .github/workflows/openjdk_14_build.yml:10]: this is the container that circleci was using, I'll try the new one

File: .github/workflows/openjdk_14_build.yml (hunk 4 above)
Comment (statop) [reply .github/workflows/openjdk_14_build.yml:19]: the gcr.io/swirlds-registry/cci-openjdk-ghr:14.0.2 container has an older version of git that doesn't work with the git checkout action.

File: .github/workflows/openjdk_14_build.yml (hunk 5 above)
Comment (statop) [reply .github/workflows/openjdk_14_build.yml:22]: sonar and the git maven plugin want to see the whole history.

Comment (steven-sheehy) [reply .github/workflows/openjdk_14_build.yml:22]: Ah, right Sonar recommends that.

File: .github/workflows/openjdk_14_build.yml (hunk 10 above)
Comment (statop) [reply .github/workflows/openjdk_14_build.yml:37]: I'll add a backlog item to look into this.

Comment (statop) [reply .github/workflows/openjdk_14_build.yml:37]: https://github.com/hashgraph/hedera-transaction-tool-demo/issues/258

Comment (statop) [reply .github/workflows/openjdk_14_build.yml:37]: nevermind, doing this

File: .github/workflows/openjdk_14_build.yml [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+          distribution: 'adopt'
+          java-version: 14
+      - name: Cache SonarCloud packages
+        uses: actions/cache@v1
+        with:
+          path: ~/.sonar/cache
+          key: ${{ runner.os }}-sonar
+          restore-keys: ${{ runner.os }}-sonar
+      - name: Compile
+        run: ./mvnw --batch-mode --no-transfer-progress --show-version -DskipTests clean compile
```
[the remark this answers is not in the document]
Comment (statop) [reply .github/workflows/openjdk_14_build.yml:28]: I can see wanting to have interactive mode and transfer progress show locally?

[the remark this answers is not in the document]
Comment (steven-sheehy) [reply .github/workflows/openjdk_14_build.yml]: I couldn't, but to each their own. For sure `--show-version` could move to `maven.config`. And the others could be at least moved to a `MAVEN_CLI_OPTS` env variable to stay dry.

**Review by steven-sheehy — APPROVED** [review]
LGTM

File: .github/workflows/openjdk_14_build.yml [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+  pull_request:
+  push:
+    branches: [ main, develop ]
+    tags: [ v* ]
+jobs:
+  build-and-test:
+    runs-on: ubuntu-latest
+    timeout-minutes: 60
+    env:
+      MAVEN_CLI_OPTS: --batch-mode --no-transfer-progress
```
Comment (steven-sheehy) [inline .github/workflows/openjdk_14_build.yml:12]: I believe you still need to put `./mvnw $MAVEN_CLI_OPTS` for this to work. There's no built-in environment variable that maven uses for extra flags.

Comment (statop) [reply .github/workflows/openjdk_14_build.yml:12]: doh, fixing
